File: ios/Classes/RxFlutterMethodChannel.swift
import Flutter
import RxSwift

public class RxFlutterPluginMethodChannel {
    let methodChannel: FlutterMethodChannel
    
    var cachedDisposables: [Int: Disposable] = [:]
    var storedObservables: [String: ObservableSourceHolder] = [:]
    
    public init(
        _ channelName: String,
        _ binaryMessenger: FlutterBinaryMessenger
        ) {
        self.methodChannel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: binaryMessenger
        )
        
        methodChannel.setMethodCallHandler{ (methodCall: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            RxFlutterPluginLogger.d("onMethodCall = channel: \(channelName), method: \(methodCall.method), arguments: \(methodCall.arguments)")
            
            //Attempt to cast to ObservableRegistrationRequest
            do {
                let observableRegistration = try ObservableRegistrationRequest(methodCall)
                switch observableRegistration.observableAction {
                case ObservableAction.SUBSCRIBE:
                    let observableSource = self.storedObservables[observableRegistration.method]
                    if observableSource == nil
                    {
                        result(
                            ObservableRegistrationResponse(
                                errorCode: PluginError.getErrorCode(PluginError.ObservableNotAvailable(nil)),
                                errorMessage: "Could not find source for \(observableRegistration.method)"
                                ).toDictionary()
                        )
                    }
                    else
                    {
                        if observableRegistration.streamType != observableSource!.type {
                            result(
                                ObservableRegistrationResponse(
                                    errorCode: PluginError.getErrorCode(PluginError.InvalidObservableType(nil)),
                                    errorMessage: "Expected \(observableRegistration.streamType.rawValue) but found \(observableSource!.type.rawValue)"
                                    )
                                    .toDictionary()
                            )
                        }
                        else
                        {
                            do
                            {
                                self.subscribeToSource(
                                    requestId: observableRegistration.requestId,
                                    source: observableSource!,
                                    arguments: observableRegistration.arguments
                                )
                                result(
                                    ObservableRegistrationResponse(
                                        errorCode: 0,
                                        errorMessage: nil
                                        )
                                        .toDictionary()
                                )
                            }
                            catch
                            {
                                result(
                                    ObservableRegistrationResponse(
                                        errorCode: PluginError.getErrorCode(PluginError.ObservableState(nil)),
                                        errorMessage: error.localizedDescription
                                        )
                                        .toDictionary()
                                )
                            }
                        }
                    }
                    
                case ObservableAction.DISPOSE:
                    let cachedDisposable = self.cachedDisposables.removeValue(forKey: observableRegistration.requestId)
                    if cachedDisposable == nil {
                        RxFlutterPluginLogger.w("Could not find subscribed observable for requestId \(observableRegistration.requestId)}")
                        result(
                            ObservableRegistrationResponse(
                                errorCode: 0,
                                errorMessage: nil
                                )
                                .toDictionary()
                        )
                    }
                    else
                    {
                        cachedDisposable?.dispose()
                        result(
                            ObservableRegistrationResponse(
                                errorCode: 0,
                                errorMessage: nil
                                )
                                .toDictionary()
                        )
                    }
                }
            } catch {
                RxFlutterPluginLogger.e("Encountered an error trying to handle methodCall.", error)
            }
        }
    }
    
    public func addObservable(
        _ method: String,
        _ sourceGen: @escaping (Any?) throws -> Observable<Any>,
        _ errorHandler: ((Error) -> Any?)?
        ) {
        let existingObservable = storedObservables.removeValue(forKey: method)
        if existingObservable != nil {
            RxFlutterPluginLogger.w("Removing existing observable source for method: \(method)")
        }
        storedObservables[method] = ObservableSourceHolderImpl(
            observable: sourceGen,
            errorHandler
            )
        RxFlutterPluginLogger.d("Added observable source for method: \(method)")
    }
    
    public func addSingle(
        _ method: String,
        _ sourceGen: @escaping (Any?) throws -> PrimitiveSequence<SingleTrait, Any>,
        _ errorHandler: ((Error) -> Any?)?
        ) {
        let existingObservable = storedObservables.removeValue(forKey: method)
        if existingObservable != nil {
            RxFlutterPluginLogger.w("Removing existing single source for method: \(method)")
        }
        storedObservables[method] = ObservableSourceHolderImpl(
            single: sourceGen,
            errorHandler
            )
        RxFlutterPluginLogger.d("Added single source for method: \(method)")
    }
    
    public func addCompletable(
        _ method: String,
        _ sourceGen: @escaping (Any?) throws -> PrimitiveSequence<CompletableTrait, Never>,
        _ errorHandler: ((Error) -> Any?)?
        ) {
        let existingObservable = storedObservables.removeValue(forKey: method)
        if existingObservable != nil {
            RxFlutterPluginLogger.w("Removing existing completable source for method: \(method)")
        }
        storedObservables[method] = ObservableSourceHolderImpl(
            completable: sourceGen,
            errorHandler
            )
        RxFlutterPluginLogger.d("Added completable source for method: \(method)")
    }
    
    private func subscribeToSource(
        requestId: Int,
        source: ObservableSourceHolder,
        arguments: Any?
        )
    {
        switch source.type {
        case StreamType.OBSERVABLE:
            cachedDisposables[requestId] = Observable.deferred {
                return try source.getSourceAsObservable(arguments)
                }
                .subscribe(
                    onNext: { element in
                        RxFlutterPluginLogger.d("onNext = requestId: \(requestId), payload: \(element)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_NEXT,
                                requestId: requestId,
                                payload: element,
                                errorMessage: nil
                            )
                        )
                },
                    onError: { error in
                        RxFlutterPluginLogger.d("onError = requestId: \(requestId), error: \(error)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_ERROR,
                                requestId: requestId,
                                payload: source.errorHandler?(error),
                                errorMessage: error.localizedDescription
                            )
                        )
                },
                    onCompleted: {
                        RxFlutterPluginLogger.d("onCompleted = requestId: \(requestId)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_COMPLETE,
                                requestId: requestId,
                                payload: nil,
                                errorMessage: nil
                            )
                        )
                }
            )
        case StreamType.SINGLE:
            cachedDisposables[requestId] = Single.deferred {
                return try source.getSourceAsSingle(arguments)
                }
                .subscribe(
                    onSuccess: { element in
                        RxFlutterPluginLogger.d("onSuccess = requestId: \(requestId), payload: \(element)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_NEXT,
                                requestId: requestId,
                                payload: element,
                                errorMessage: nil
                            )
                        )
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_COMPLETE,
                                requestId: requestId,
                                payload: nil,
                                errorMessage: nil
                            )
                        )
                },
                    onError: { error in
                        RxFlutterPluginLogger.d("onError = requestId: \(requestId), error: \(error)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_ERROR,
                                requestId: requestId,
                                payload: source.errorHandler?(error),
                                errorMessage: error.localizedDescription
                            )
                        )
                }
            )
        case StreamType.COMPLETABLE:
            cachedDisposables[requestId] = Completable.deferred {
                return try source.getSourceAsCompletable(arguments)
                }
                .subscribe(
                    onCompleted: {
                        RxFlutterPluginLogger.d("onCompleted = requestId: \(requestId)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_COMPLETE,
                                requestId: requestId,
                                payload: nil,
                                errorMessage: nil
                            )
                        )
                },
                    onError: { error in
                        RxFlutterPluginLogger.d("onError = requestId: \(requestId), error: \(error)")
                        self.sendObservableCallback(
                            ObservableCallback(
                                callbackType: ObservableCallback.CallbackType.ON_ERROR,
                                requestId: requestId,
                                payload: source.errorHandler?(error),
                                errorMessage: error.localizedDescription
                            )
                        )
                }
            )
        }
    }
    
    private func sendObservableCallback(_ observableCallback: ObservableCallback) {
        RxFlutterPluginLogger.d("sendObservableCallback(): \(observableCallback)")
        DispatchQueue.main.async {
            self.methodChannel.invokeMethod(
                "callback",
                arguments: NSDictionary(dictionary: [
                    Field.OBSERVABLE_CALLBACK.rawValue: observableCallback.callbackType.rawValue,
                    Field.PAYLOAD.rawValue: observableCallback.payload,
                    Field.REQUEST_ID.rawValue: observableCallback.requestId,
                    Field.ERROR_MESSAGE.rawValue: observableCallback.errorMessage
                ])
            )
        }
    }
}

File: android/src/main/kotlin/tenko/rx_flutter_plugin/RxFlutterMethodChannel.kt
package tenko.rx_flutter_plugin

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.reactivex.*
import io.reactivex.disposables.Disposable

class RxFlutterMethodChannel(val channelName: String, binaryMessenger: BinaryMessenger) {
    private val methodChannel = MethodChannel(binaryMessenger, channelName)

    private val methodChannelCallback: (MethodCall, MethodChannel.Result) -> Unit = { methodCall, result ->
        Logger.d("MethodChannelCallback = channel: $channelName, method: ${methodCall.method}, args: ${methodCall.arguments}")
        val observableRegistration = ObservableRegistrationRequest.constructFromMethodCall(methodCall)
        when (observableRegistration.observableAction) {
            ObservableAction.subscribe -> {
                val observableSource = storedObservables[observableRegistration.method]
                if (observableSource == null) {
                    result.success(
                            ObservableRegistrationResponse(
                                    PluginException.ObservableNotAvailable().errorCode,
                                    "Could not find source for ${observableRegistration.method}"
                            )
                                    .toHashMap()
                    )
                }
                else {
                    //Do some verification
                    when {
                        observableRegistration.streamType != observableSource.type -> {
                            result.success(
                                    ObservableRegistrationResponse(
                                            PluginException.InvalidObservableType().errorCode,
                                            "Expected ${observableRegistration.streamType} but found ${observableSource.type} for method: ${observableRegistration.method}."
                                    )
                                            .toHashMap()
                            )

                        }

                        else -> {
                            try {
                                subscribeToSource(
                                        observableRegistration.requestId,
                                        observableSource,
                                        methodCall.argument(Field.PAYLOAD)
                                )
                                result.success(
                                        ObservableRegistrationResponse(
                                                0
                                        )
                                                .toHashMap()
                                )
                            } catch (e: Exception) {
                                result.success(
                                        ObservableRegistrationResponse(
                                                PluginException.ObservableThrown().errorCode,
                                                e.localizedMessage
                                        )
                                                .toHashMap()
                                )
                            }
                        }
                    }
                }
            }

            ObservableAction.dispose -> {
                val cachedDisposable = cachedDisposables[observableRegistration.requestId]
                if (cachedDisposable == null) {
                    Logger.w("Could not find subscribed observable for requestId ${observableRegistration.requestId}.")
                    result.success(
                            ObservableRegistrationResponse(
                                    0
                            )
                                    .toHashMap()
                    )
                } else {
                    cachedDisposable.dispose()
                    cachedDisposables.remove(observableRegistration.requestId)
                    result.success(
                            ObservableRegistrationResponse(
                                    0
                            )
                                    .toHashMap()
                    )
                }
            }
        }
    }

    init {
        methodChannel.setMethodCallHandler(methodChannelCallback)
    }

    /**
     * Used to hold the mappings of methods to the observable source.
     */
    private val storedObservables = HashMap<String, ObservableSourceHolder<*>>()

    fun <T> addObservable(
            methodName: String,
            source: (arguments: T?) -> ObservableSource<Any>,
            errorHandler: ((Throwable) -> Any?)? = null
    ) {
        Logger.d("Observable source registered. methodName: $methodName")
        storedObservables[methodName] = ObservableSourceHolder.observableHolder(source, errorHandler)
    }

    fun <T> addSingle(
            methodName: String,
            source: (arguments: T?) -> SingleSource<Any>,
            errorHandler: ((Throwable) -> Any?)? = null
    ) {
        Logger.d("Single source registered. methodName: $methodName")
        storedObservables[methodName] = ObservableSourceHolder.singleHolder(source, errorHandler)
    }

    fun <T> addCompletable(
            methodName: String,
            source: (arguments: T?) -> CompletableSource,
            errorHandler: ((Throwable) -> Any?)? = null
    ) {
        Logger.d("Completable source registered. methodName: $methodName")
        storedObservables[methodName] = ObservableSourceHolder.completableHolder(source, errorHandler)
    }

    private val mainThreadHandler = Handler(Looper.getMainLooper())
    private fun sendObservableCallback(observableCallback: ObservableCallback) {
        Logger.d("sendObservableCallback(): $observableCallback")
        //Use arbitrary method for now, method is not used.
        //No need to include callback. It bears no useful information
        //Ensure all calls are done on UI thread
        mainThreadHandler.post{
            methodChannel.invokeMethod(
                    "callback",
                    mapOf(
                            Field.OBSERVABLE_CALLBACK to observableCallback.observableCallbackType.name,
                            Field.PAYLOAD to observableCallback.payload,
                            Field.REQUEST_ID to observableCallback.requestId,
                            Field.ERROR_MESSAGE to observableCallback.errorMessage
                    )
            )
        }
    }


    private val cachedDisposables = HashMap<Int, Disposable>()
    private fun <T> subscribeToSource(requestId: Int, source: ObservableSourceHolder<T>, arguments: T?) {
        cachedDisposables[requestId] =
                when (source.type) {
                    StreamType.observable -> {
                        Observable.defer {
                            source.getSourceAsObservable(arguments)
                        }.subscribe({
                            Logger.d("onNext() = requestId: $requestId, payload: $it")
                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onNext,
                                            requestId,
                                            it,
                                            null
                                    )
                            )
                        }, {
                            Logger.w(it, "onError() = requestId: $requestId")

                            val payload =
                                    when {
                                        source.errorHandler != null -> source.errorHandler?.invoke(it)
                                        defaultExceptionHandler != null -> defaultExceptionHandler?.handleException(it)
                                        else -> null
                                    }

                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onError,
                                            requestId,
                                            payload,
                                            it.localizedMessage
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        }, {
                            Logger.d("onComplete() = requestId: $requestId")
                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onComplete,
                                            requestId,
                                            null,
                                            null
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        })
                    }

                    StreamType.single -> {
                        Single.defer {
                            source.getSourceAsSingle(arguments)
                        }.subscribe({
                            Logger.d("onSuccess() = requestId: $requestId, payload: $it")
                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onNext,
                                            requestId,
                                            it,
                                            null
                                    )
                            )
                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onComplete,
                                            requestId,
                                            null,
                                            null
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        }, {
                            Logger.w(it, "onError() = requestId: $requestId")

                            val payload =
                                    when {
                                        source.errorHandler != null -> source.errorHandler?.invoke(it)
                                        defaultExceptionHandler != null -> defaultExceptionHandler?.handleException(it)
                                        else -> null
                                    }

                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onError,
                                            requestId,
                                            payload,
                                            it.localizedMessage
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        })
                    }

                    StreamType.completable -> {
                        Completable.defer {
                            source.getSourceAsCompletable(arguments)
                        }.subscribe({
                            Logger.d("onComplete() = requestId: $requestId")
                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onComplete,
                                            requestId,
                                            null,
                                            null
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        }, {
                            Logger.w(it, "onError() = requestId: $requestId")

                            val payload =
                                    when {
                                        source.errorHandler != null -> source.errorHandler?.invoke(it)
                                        defaultExceptionHandler != null -> defaultExceptionHandler?.handleException(it)
                                        else -> null
                                    }

                            sendObservableCallback(
                                    ObservableCallback(
                                            ObservableCallbackType.onError,
                                            requestId,
                                            payload,
                                            it.localizedMessage
                                    )
                            )
                            cachedDisposables.remove(requestId)
                        })
                    }
                }
    }

    private var defaultExceptionHandler: ExceptionHandler? = null

    /**
     * Set the default ExceptionHandler to be used when an Rx subscription throws.
     * If null is returned by the given ExceptionHandler, the default behaviour is used (ObservableThrownException).
     */
    fun setDefaultExceptionHandler(exceptionHandler: ExceptionHandler) {
        defaultExceptionHandler = exceptionHandler
    }
}